[Mau Ze Wei] iP

wassilililily opened the pull request:
# BottleOpener

> "I can't believe you lost our bottle opener." - J.D. 

Don't worry about losing your bottle opener again (or anything else for that matter) as BottleOpener will
* help you keep track of tasks
* search through your list of tasks quickly

All you need to do is:
1. Download it from [here](https://github.com/wassilililily/ip/releases/tag/A-Jar)
2. Run it
3.  **Profit?!?** 👍 

Some of the basic commands are:
```
bye
list
delete
todo
deadline
event
```

You can even do this:
- [ ] mark
- [ ] unmark

Here is a snippet of the code:

```ruby
public class BottleOpener {
    public static void main(String[] args) {
        String botName = "BottleOpener";
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wassilililily committed:
Renamed main class

wassilililily committed:
Added goodbye message

wassilililily committed:
Changed new line command and added more dash

wassilililily committed:
Moved punctuation

wassilililily committed:
Added echo

wassilililily committed:
Modified greeting

wassilililily committed:
Add list

wassilililily committed:
Added task class

wassilililily committed:
Added ability to mark/unmark and modified Task class

wassilililily committed:
Added todo and deadline, change if-else to switch

wassilililily committed:
Added event

wassilililily committed:
Added automated ui testing

wassilililily committed:
Added delete

wassilililily committed:
Add ability to save

wassilililily committed:
Merge branch 'branch-level-7'

wassilililily committed:
Add ability to save date and time

wassilililily committed:
Merge branch 'branch-level-8'

wassilililily committed:
Add Parser, Storage, Tasklist and Ui classes

wassilililily committed:
Organise classes into one package

wassilililily committed:
Merge branch 'master' into add-gradle-support

wassilililily committed:
Merge branch 'branch-increments'

wassilililily committed:
Merge branch 'add-gradle-support'

wassilililily committed:
Add JUnit testing

wassilililily committed:
Add find function

hanscau commented on src/main/java/BottleOpener/BottleOpener.java:
```diff
@@ -0,0 +1,47 @@
+     * @param args Command-line arguments (not used).
+     */
+    public static void main(String[] args) {
+        String botName = "BottleOpener";
+        String filepath = "data/BottleOpener.txt";
```
Should change these into a constant instead since they are hard coded
```java
private final String BOT_NAME = "BottleOpener";
private final String FILE_PATH = "data/BottleOpener.txt";
```

hanscau commented on src/main/java/BottleOpener/Deadline.java:
```diff
@@ -0,0 +1,134 @@
+    public Deadline(String description, String deadline) {
+        super(description);
+        try {
+            this.deadline = LocalDateTime.parse(deadline, FORMATTER);
```
Very cool to allow different formats of datetime 👍

hanscau commented on src/main/java/BottleOpener/Event.java:
```diff
@@ -0,0 +1,161 @@
+public class Event extends Task {
+    private static final DateTimeFormatter FORMATTER;
+
+    static {
+        FORMATTER = new DateTimeFormatterBuilder()
+                .appendOptional(DateTimeFormatter.ofPattern("dd MMM yyyy HHmm"))
+                .appendOptional(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm"))
+                .appendOptional(DateTimeFormatter.ofPattern("dd MMM yyyy"))
+                .appendOptional(DateTimeFormatter.ofPattern("dd/M/yyyy HHmm"))
+                .appendOptional(DateTimeFormatter.ofPattern("dd/M/yyyy HH:mm"))
+                .appendOptional(DateTimeFormatter.ofPattern("dd/M/yyyy"))
+                .appendOptional(DateTimeFormatter.ofPattern("d/M/yyyy"))
+                .appendOptional(DateTimeFormatter.ofPattern("d/M/yyyy HH:mm"))
+                .appendOptional(DateTimeFormatter.ofPattern("d/M/yyyy HHmm"))
+                .appendOptional(DateTimeFormatter.ofPattern("MM/dd/yyyy HHmm"))
+                .appendOptional(DateTimeFormatter.ofPattern("MM/dd/yyyy HH:mm"))
+                .appendOptional(DateTimeFormatter.ofPattern("MM/dd/yyyy"))
+                .appendOptional(DateTimeFormatter.ofPattern("MM-dd-yyyy HHmm"))
+                .appendOptional(DateTimeFormatter.ofPattern("MM-dd-yyyy HH:mm"))
+                .appendOptional(DateTimeFormatter.ofPattern("MM-dd-yyyy"))
+                .parseDefaulting(ChronoField.HOUR_OF_DAY, 0)
+                .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 0)
+                .toFormatter();
+    }
```
Duplicated code, maybe can abstract them away? or just have one of the class import it from the other class

hanscau commented on src/main/java/BottleOpener/Event.java:
```diff
@@ -0,0 +1,161 @@
+        try {
+            this.start = LocalDateTime.parse(start, FORMATTER);
+            this.end = LocalDateTime.parse(end, FORMATTER);
+        } catch (DateTimeParseException e) {
+            System.out.println("Invalid event format!");
+            this.start = LocalDateTime.now();
+            this.end = LocalDateTime.now();
+        }
```
nit: does not seem right to force the event to have current time for start and end if the format is wrong

hanscau commented on src/main/java/BottleOpener/Storage.java:
```diff
@@ -0,0 +1,85 @@
+                        String type = word[0];
+                        boolean status = word[1].contains("X");
+
+                        if (type.equals("T")) {
+                            Task newTask = new ToDo(word[2], status);
+                            tasks.addTask(newTask);
+                        } else if (type.equals("D")) {
+                            Task newTask = new Deadline(word[2], status, word[3]);
+                            tasks.addTask(newTask);
+                        } else if (type.equals("E")) {
```
nit: can call `Task.getType()`, `Deadline.getType()`, and `Event.getType()` instead so there is no need to hardcode it in and since you have them defined in the individual classes already

hanscau reviewed (changes requested):
Overall very good and clean code writing. Very verbose javadocs and love the fact that the events are treated as if they are immutable. 

Just a few suggestions and weird behaviour that I don't think make sense for the user

PresleyChew commented on src/main/java/BottleOpener/BottleOpener.java:
```diff
@@ -0,0 +1,47 @@
+
+        Tasklist tasks = new Tasklist();
+        tasks = storage.load(tasks);
+        boolean flag = false;
```
Perhaps rename "flag" to something else, as booleans should sound like booleans, but overall code is amazing! 😸

PresleyChew commented on src/main/java/BottleOpener/BottleOpener.java:
```diff
@@ -0,0 +1,47 @@
+        boolean flag = false;
+        while (!flag) {
+            storage.save(tasks);
+            String inp = "";
```
Perhaps consider renaming "inp" to something more clear!

PresleyChew commented on src/main/java/BottleOpener/ToDo.java:
```diff
@@ -0,0 +1,57 @@
+        return new ToDo(this.description, true);
+    }
+
+    public ToDo markAsUndone() {
```
Great code naming! Using verbs for actions, great job ! 🔥

PresleyChew commented on src/main/java/BottleOpener/Tasklist.java:
```diff
@@ -0,0 +1,105 @@
+ */
+public class Tasklist {
+    public int index;
+    public ArrayList<Task> tasklist;
```
Perhaps consider renaming it to taskLists, following camelCasing as well as distinguishing clearly between single-valued and multi-valued variables! Small nit picking!

PresleyChew reviewed (commented):
LGTM with minor nit picking with naming, otherwise code is great! 💯

wassilililily committed:
Add new Util class and change attributes to private

wassilililily committed:
Refactor parser attributes and parsertest date format

wassilililily committed:
Add checkstyle and change package name

wassilililily committed:
Add gui

wassilililily committed:
Add new line on launcher

wassilililily committed:
Merge branch 'branch-level-10'

wassilililily committed:
Add assertions

wassilililily committed:
Add command package and refactor parser and Ui class

wassilililily committed:
Merge pull request #1 from wassilililily/branch-A-Assertions

Add assertions

wassilililily committed:
Merge branch 'master' of https://github.com/wassilililily/ip

wassilililily committed:
Edit master branch to resolve merge conflicts

wassilililily committed:
Delete unnecessary import

wassilililily committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/bottleopener/myapp/BottleOpener.java
#	src/main/java/bottleopener/myapp/Parser.java
#	src/main/java/bottleopener/ui/Ui.java

wassilililily committed:
Merge pull request #2 from wassilililily/branch-A-CodeQuality

Add command package and refactor parser and Ui class

wassilililily committed:
Merge branch 'master' of https://github.com/wassilililily/ip

wassilililily committed:
Merge branch 'master' of https://github.com/wassilililily/ip

wassilililily committed:
Merge pull request #3 from wassilililily/branch-A-Assertions

Branch A-Assertions

wassilililily committed:
Merge branch 'branch-level-10' of https://github.com/wassilililily/ip into branch-level-10

wassilililily committed:
Merge branch 'branch-level-10'

wassilililily committed:
Refactor Parser Class

Refactor parser class to improve code organization and maintainability
by delegating command execution to separate command classes.
The runParser() method has been broken down into distinct command
handling classes, each responsible for executing specific commands.

wassilililily committed:
Merge pull request #4 from wassilililily/branch-A-CodeQuality

Refactor Parser Class

wassilililily committed:
Delete unimportant assertions

wassilililily committed:
Merge commit 'ea960cac9c881ffbbde459fa2ede4c8a0fb4e14e' into branch-A-Assertions

# Conflicts:
#	src/main/java/bottleopener/myapp/BottleOpener.java
#	src/main/java/bottleopener/myapp/Parser.java

wassilililily committed:
Merge pull request #5 from wassilililily/branch-A-Assertions

Delete unimportant assertions

wassilililily committed:
Merge branch 'master' of https://github.com/wassilililily/ip

wassilililily committed:
Add ErrorCommand and RemindCommand class

Refactor parser class so that it runs the runCommand() method
separately. The ErrorCommand is the default in the switch-case
block.

Add RemindCommand to implement the remind extension, reminding
user of their deadlines during initilialisation and the remind command.

wassilililily committed:
Merge branch 'branch-Extension'

wassilililily committed:
Add user guide

wassilililily committed:
Add screenshot of GUI

wassilililily committed:
Update user guide

wassilililily committed:
Modify boolean variable name

wassilililily committed:
Add JUnit test for parser and modify assertion for task list

Parser class currently lacks automated tests, making it difficult to
verify the correctness of command parsing.

Testing coverage needs to be added to ensure that the Parser class
correctly processes all supported commands.

Introduce a JUnit test suite that covers valid commands (such as 'list',
'bye', 'mark', 'unmark', 'delete', 'remind', 'todo', 'deadline', and 'find'),
and add checks for command error.

These tests are written to ensure thorough coverage of the Parser's
behavior, making it easier to detect issues during future changes.
Command parsing tests verify that the correct command is returned
based on user input.

This approach increases the maintainability and reliability of the Parser
class, and ensures the system behaves as expected under various
scenarios.
